src/main/java/hudson/tasks/junit/CaseResult.java

```diff
@@ -256,10 +262,25 @@ private static String getSkippedMessage(Element testCase) {
         return message;
     }
 
-    public String getDisplayName() {
+    public String getTransformedTestName() {
         return TestNameTransformer.getTransformedName(testName);
     }
 
+    public String getDisplayName() {
+        // Only prepend the enclosing flow node names if there are any and the run this is in has multiple blocks directly containing
+        // test results.
+        if (!getEnclosingFlowNodeNames().isEmpty()) {
```

Comment:
Excellent!  Yes, that's very close to what I was thinking - with just a couple tweaks: 

1. We don't need to do a scan for every new node.  We can just use the mapping for the previous node... after checking if it is itself a block start/end.  If it's a block end, then we get the block start node corresponding to *that* and examine it.  Yay presto, enclosing block in O(1) time for the common case, with a tiny handful of cases where you need to walk further up.
2.  Probably worth having an early exit for FlowStartNode
3. Maybe we want to use String IDs entirely rather than FlowNode objects, when possible?

I doubt any of those actually change the logic materially that much though.  What you've got is more or less exactly what I was thinking!  :man_dancing: 

One note: we may want to ensure the actual implementation itself is never exposed in case (cough) this is supported directly via the FlowNodeStorage implementations in the future.